File: conformance/tests/p256-wallet/p256-wallet.runar.py
from runar import SmartContract, ByteString, Addr, Sig, PubKey, public, assert_, hash160, check_sig, verify_ecdsa_p256

class P256Wallet(SmartContract):
    ecdsa_pub_key_hash: Addr
    p256_pub_key_hash: ByteString

    def __init__(self, ecdsa_pub_key_hash: Addr, p256_pub_key_hash: ByteString):
        super().__init__(ecdsa_pub_key_hash, p256_pub_key_hash)
        self.ecdsa_pub_key_hash = ecdsa_pub_key_hash
        self.p256_pub_key_hash = p256_pub_key_hash

    @public
    def spend(self, p256_sig: ByteString, p256_pub_key: ByteString, sig: Sig, pub_key: PubKey):
        assert_(hash160(pub_key) == self.ecdsa_pub_key_hash)
        assert_(check_sig(sig, pub_key))
        assert_(hash160(p256_pub_key) == self.p256_pub_key_hash)
        assert_(verify_ecdsa_p256(sig, p256_sig, p256_pub_key))

File: conformance/tests/p256-wallet/p256-wallet.runar.ts
import { SmartContract, assert, checkSig, hash160, verifyECDSA_P256 } from 'runar-lang';
import type { ByteString, Addr, Sig, PubKey } from 'runar-lang';

class P256Wallet extends SmartContract {
  readonly ecdsaPubKeyHash: Addr;
  readonly p256PubKeyHash: ByteString;

  constructor(ecdsaPubKeyHash: Addr, p256PubKeyHash: ByteString) {
    super(ecdsaPubKeyHash, p256PubKeyHash);
    this.ecdsaPubKeyHash = ecdsaPubKeyHash;
    this.p256PubKeyHash = p256PubKeyHash;
  }

  public spend(p256Sig: ByteString, p256PubKey: ByteString, sig: Sig, pubKey: PubKey) {
    assert(hash160(pubKey) == this.ecdsaPubKeyHash);
    assert(checkSig(sig, pubKey));
    assert(hash160(p256PubKey) == this.p256PubKeyHash);
    assert(verifyECDSA_P256(sig, p256Sig, p256PubKey));
  }
}
